The declared narwhals floor (>= 1.0.0) is untested, and the real one conflicts with requires-python = ">= 3.8"

Found while porting `_utils.convert_df_scientific` for #37. Not a rewrite problem — it predates the chain — so raising it rather than folding a fix into a slice.

## The declared floor is not the real one

`pyproject.toml` said `narwhals >= 1.0.0`. Checked out `main` (before any slice) and ran the suite against a range of narwhals versions:

| narwhals | `uv run pytest` on `main` |
|---|---|
| 1.40.0 | **5 failed**, 46 passed |
| 1.45.0 | **5 failed**, 46 passed |
| 2.0.0 | 51 passed |
| 2.10.0 | 51 passed |
| 2.20.0 | 51 passed |

The five are all the pandas backend:

```
FAILED tests/test_backends.py::test_pandas_backend_basic - assert 3 == 2
FAILED tests/test_backends.py::test_polars_vs_pandas_categorical_stats
FAILED tests/test_backends.py::test_backend_with_empty_numeric
FAILED tests/test_backends.py::test_pandas_backend_categorical_top_values
FAILED tests/test_table.py::test_pandas - assert (1, 7) == (2, 7)
```

Before 2.0.0, narwhals types pandas `object` columns differently, so string columns are classified as numerical and land in the wrong table. So the true floor for a working pandas backend is **narwhals >= 2.0.0**.

## Why it cannot simply be bumped

narwhals 2.0.0 requires Python >= 3.9; 2.21.2+ requires >= 3.10. `requires-python` here is `>= 3.8`, and `uv sync` refuses the combination outright:

```
× No solution found when resolving dependencies for split (markers: python_full_version == '3.8.*'):
  ╰─▶ Because the requested Python version (>=3.8) does not satisfy Python>=3.9
      and narwhals>=2.0.0,<=2.21.0 depends on Python>=3.9 …
  hint: Consider using a more restrictive `requires-python` value (like >=3.9).
```

So `narwhals >= 2.0.0` and `requires-python = ">= 3.8"` cannot both hold. One of them has to give, and which one is a maintainer call:

1. **Drop 3.8** — `requires-python = ">= 3.9"`, `narwhals >= 2.0.0`, drop the 3.8 leg from `noxfile.py`. 3.8 has been end-of-life since October 2024.
2. **Keep 3.8** — then the pandas backend is only supported on 3.9+ in practice, and the 3.8 leg of `noxfile.py` is expected to fail; that should at least be written down.

There is a related smell either way: the locked narwhals is 2.24.0, which needs Python >= 3.10. So `uv run nox`'s 3.8 and 3.9 legs resolve to a narwhals nothing has ever been tested against — which is presumably why the failures above went unnoticed.

## What I did in the meantime

The #37 slice that touches `_utils` uses `Expr.log`, which arrived in narwhals **1.40.0**, so the floor moved `1.0.0 → 1.40.0` there. That is the honest minimum for the code as written, and it resolves under 3.8. It deliberately does **not** decide the question above.

Two things kept the floor that low rather than following the installed 2.24:

- `Expr.floor` only exists from 2.20, so the port uses `// 1`, which floors identically on polars, pandas and pyarrow.
- Chaining `.when()` onto a `Then` only works from 2.x, so the port nests `when/then/otherwise` instead.

Both are noted in the code. If the decision is to drop 3.8, they can be simplified.

--- a/.github/workflows/pytest.yaml
+++ b/.github/workflows/pytest.yaml
@@ -22,3 +22,9 @@ jobs:
 
       - name: Run tests
         run: uv run pytest -v -m "not integration"
+
+      - name: Install minimum supported Python
+        run: uv python install 3.10
+
+      - name: Test minimum supported Narwhals
+        run: uv run nox -s minimum_versions

--- a/changelog.md
+++ b/changelog.md
@@ -36,6 +36,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Changed
 
+- CI now runs the stable suite with Python 3.10 and Narwhals 2.20.0, so the
+  declared minimum versions are checked on every PR. The pandas extra now
+  requires PyArrow 13, which is the minimum supported by Narwhals 2.20.0
+  (#78)
 - **Breaking:** `requires-python` is now `>= 3.10`, up from `>= 3.8`, and the
   minimum `narwhals` is 2.20.0. The two are linked: `requires-python` caps
   which narwhals is installable at all — 1.42.1 and below need 3.8, 1.43.0

--- a/noxfile.py
+++ b/noxfile.py
@@ -23,11 +23,28 @@ def test(session):
         "hypothesis>=6.113.0",
         "polars>=0.20.21",
         "pandas>=1.5.3",
-        "pyarrow>=10.0.0",
+        "pyarrow>=13.0.0",
         "narwhals>=2.20.0",
     )
 
     session.run("pytest", "tests/")
+
+
+@nox.session(name="minimum_versions", python="3.10")
+def test_minimum_versions(session):
+    """Run the stable suite at the Python and narwhals floors."""
+    session.install(
+        "pytest==8.3.2",
+        "hypothesis>=6.113.0",
+        "polars>=0.20.21",
+        "pandas>=1.5.3",
+        "pyarrow>=13.0.0",
+        "narwhals==2.20.0",
+    )
+    # Property tests check statistical behavior against random data. They do
+    # not exercise a separate Narwhals API surface, so the ordinary CI job
+    # runs them once with the locked environment.
+    session.run("pytest", "tests/", "--ignore=tests/test_properties.py")
 
 
 @nox.parametrize("polars_version", ["0.20.21", "1.4.1"])
@@ -39,7 +56,7 @@ def test_polars_versions(session, polars_version, pandas_version):
         "hypothesis>=6.113.0",
         f"polars=={polars_version}",
         f"pandas>={pandas_version}",
-        "pyarrow>=10.0.0",
+        "pyarrow>=13.0.0",
         "narwhals>=2.20.0",
     )
     session.run("pytest", "tests/")

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -29,7 +29,7 @@ requires-python = ">= 3.10"
 # Convenience extras for people who want a dataframe library installed
 # alongside showstats. None of them is required: showstats works with
 # whatever narwhals-compatible frame it is handed.
-pandas = ["pandas>=1.5.3", "pyarrow>=10.0.0"]
+pandas = ["pandas>=1.5.3", "pyarrow>=13.0.0"]
 polars = ["polars>=0.20.21"]
 
 

